# PCOS Disease Prediction using Machine Learning

## Import Libraries

In [1]:
import pandas as pd

## Load Dataset

In [2]:
df = pd.read_csv(r"C:\Users\WELCOME\Desktop\PCOS-Disease-Prediction\processed\pcos_cleaned.csv")

## Dataset Preview

In [3]:
df.head()

,Sl. No,Patient File No.,PCOS (Y/N),Age (yrs),Weight (Kg),Height(Cm),BMI,Blood Group,Pulse rate(bpm),RR (breaths/min),...,Unnamed: 49,Unnamed: 50,Unnamed: 51,Unnamed: 52,Unnamed: 53,Unnamed: 54,Unnamed: 55,Unnamed: 56,Unnamed: 57,Unnamed: 58
0,1.0,1.0,0.0,28.0,44.6,152.0,19.30,15.0,78.0,22.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2.0,2.0,0.0,36.0,65.0,161.5,24.92,15.0,74.0,20.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,3.0,3.0,1.0,33.0,68.8,165.0,25.27,11.0,72.0,18.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,4.0,4.0,0.0,37.0,65.0,148.0,29.67,13.0,72.0,20.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,5.0,5.0,0.0,25.0,52.0,161.0,20.06,11.0,72.0,18.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## Dataset Info

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 998 entries, 0 to 997
Data columns (total 59 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Sl. No                  540 non-null    float64
 1   Patient File No.        540 non-null    float64
 2   PCOS (Y/N)              540 non-null    float64
 3    Age (yrs)              540 non-null    float64
 4   Weight (Kg)             540 non-null    float64
 5   Height(Cm)              540 non-null    float64
 6   BMI                     540 non-null    str    
 7   Blood Group             540 non-null    float64
 8   Pulse rate(bpm)         540 non-null    float64
 9   RR (breaths/min)        540 non-null    float64
 10  Hb(g/dl)                540 non-null    float64
 11  Cycle(R/I)              540 non-null    float64
 12  Cycle length(days)      540 non-null    float64
 13  Marraige Status (Yrs)   539 non-null    float64
 14  Pregnant(Y/N)           540 non-null    float64
 15  

## Missing Values

In [5]:
df.isnull().sum()

Sl. No                    458
Patient File No.          458
PCOS (Y/N)                458
 Age (yrs)                458
Weight (Kg)               458
Height(Cm)                458
BMI                       458
Blood Group               458
Pulse rate(bpm)           458
RR (breaths/min)          458
Hb(g/dl)                  458
Cycle(R/I)                458
Cycle length(days)        458
Marraige Status (Yrs)     459
Pregnant(Y/N)             458
No. of aborptions         458
  I   beta-HCG(mIU/mL)    458
II    beta-HCG(mIU/mL)    458
FSH(mIU/mL)               458
LH(mIU/mL)                458
FSH/LH                    458
Hip(inch)                 458
Waist(inch)               458
Waist:Hip Ratio           458
TSH (mIU/L)               458
AMH(ng/mL)                458
PRL(ng/mL)                458
Vit D3 (ng/mL)            458
PRG(ng/mL)                458
RBS(mg/dl)                458
Weight gain(Y/N)          458
hair growth(Y/N)          458
Skin darkening (Y/N)      458
Hair loss(

# Dimension of Dataset

In [6]:
df.shape

(998, 59)

## Remove Empty Columns

In [7]:
df = df.loc[:, ~df.columns.str.contains('^Unnamed')]

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 998 entries, 0 to 997
Data columns (total 44 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Sl. No                  540 non-null    float64
 1   Patient File No.        540 non-null    float64
 2   PCOS (Y/N)              540 non-null    float64
 3    Age (yrs)              540 non-null    float64
 4   Weight (Kg)             540 non-null    float64
 5   Height(Cm)              540 non-null    float64
 6   BMI                     540 non-null    str    
 7   Blood Group             540 non-null    float64
 8   Pulse rate(bpm)         540 non-null    float64
 9   RR (breaths/min)        540 non-null    float64
 10  Hb(g/dl)                540 non-null    float64
 11  Cycle(R/I)              540 non-null    float64
 12  Cycle length(days)      540 non-null    float64
 13  Marraige Status (Yrs)   539 non-null    float64
 14  Pregnant(Y/N)           540 non-null    float64
 15  

## Remove Empty Rows

In [9]:
df = df.dropna(how="all")

In [10]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 540 entries, 0 to 539
Data columns (total 44 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Sl. No                  540 non-null    float64
 1   Patient File No.        540 non-null    float64
 2   PCOS (Y/N)              540 non-null    float64
 3    Age (yrs)              540 non-null    float64
 4   Weight (Kg)             540 non-null    float64
 5   Height(Cm)              540 non-null    float64
 6   BMI                     540 non-null    str    
 7   Blood Group             540 non-null    float64
 8   Pulse rate(bpm)         540 non-null    float64
 9   RR (breaths/min)        540 non-null    float64
 10  Hb(g/dl)                540 non-null    float64
 11  Cycle(R/I)              540 non-null    float64
 12  Cycle length(days)      540 non-null    float64
 13  Marraige Status (Yrs)   539 non-null    float64
 14  Pregnant(Y/N)           540 non-null    float64
 15  

# Dimension of Dataset

In [11]:
df.shape

(540, 44)

## Standardize Column Names

In [12]:
df.columns = df.columns.str.strip()

In [13]:
df = df.rename(columns={
    "Sl. No": "sr_no",
    "Patient File No.": "patient_id",
    "PCOS (Y/N)": "pcos",
    "Age (yrs)": "age_yrs",
    "Weight (Kg)": "weight_kg",
    "Height(Cm)": "height_cm",
    "BMI": "bmi",
    "Blood Group": "blood_group",
    "Pulse rate(bpm)": "pulse_rate",
    "RR (breaths/min)": "respiratory_rate",
    "Hb(g/dl)": "hb",
    "Cycle(R/I)": "cycle_type",
    "Cycle length(days)": "cycle_length",
    "Marraige Status (Yrs)": "marriage_yrs",
    "Pregnant(Y/N)": "pregnant",
    "No. of abortions": "abortions",
    "I beta-HCG(mIU/mL)": "beta_hcg_1",
    "II beta-HCG(mIU/mL)": "beta_hcg_2",
    "FSH(mIU/mL)": "fsh",
    "LH(mIU/mL)": "lh",
    "FSH/LH": "fsh_lh_ratio",
    "Hip(inch)": "hip_inch",
    "Waist(inch)": "waist_inch",
    "Waist:Hip Ratio": "waist_hip_ratio",
    "TSH (mIU/L)": "tsh",
    "AMH(ng/mL)": "amh",
    "PRL(ng/mL)": "prolactin",
    "Vit D3 (ng/mL)": "vit_d3",
    "PRG(ng/mL)": "progesterone",
    "RBS(mg/dl)": "rbs",
    "Weight gain(Y/N)": "weight_gain",
    "hair growth(Y/N)": "hair_growth",
    "Skin darkening (Y/N)": "skin_darkening",
    "Hair loss(Y/N)": "hair_loss",
    "Pimples(Y/N)": "pimples",
    "Fast food (Y/N)": "fast_food",
    "Reg.Exercise(Y/N)": "exercise",
    "BP _Systolic (mmHg)": "bp_sys",
    "BP _Diastolic (mmHg)": "bp_dia",
    "Follicle No. (L)": "follicle_left",
    "Follicle No. (R)": "follicle_right",
    "Avg. F size (L) (mm)": "follicle_size_left",
    "Avg. F size (R) (mm)": "follicle_size_right",
    "Endometrium (mm)": "endometrium"
})

In [14]:
df.head()

,sr_no,patient_id,pcos,age_yrs,weight_kg,height_cm,bmi,blood_group,pulse_rate,respiratory_rate,...,pimples,fast_food,exercise,bp_sys,bp_dia,follicle_left,follicle_right,follicle_size_left,follicle_size_right,endometrium
0,1.0,1.0,0.0,28.0,44.6,152.0,19.30,15.0,78.0,22.0,...,0.0,1.0,0.0,110.0,80.0,3.0,3.0,18.0,18.0,8.5
1,2.0,2.0,0.0,36.0,65.0,161.5,24.92,15.0,74.0,20.0,...,0.0,0.0,0.0,120.0,70.0,3.0,5.0,15.0,14.0,3.7
2,3.0,3.0,1.0,33.0,68.8,165.0,25.27,11.0,72.0,18.0,...,1.0,1.0,0.0,120.0,80.0,13.0,15.0,18.0,20.0,10.0
3,4.0,4.0,0.0,37.0,65.0,148.0,29.67,13.0,72.0,20.0,...,0.0,0.0,0.0,120.0,70.0,2.0,2.0,15.0,14.0,7.5
4,5.0,5.0,0.0,25.0,52.0,161.0,20.06,11.0,72.0,18.0,...,0.0,0.0,0.0,120.0,80.0,3.0,4.0,16.0,14.0,7.0


In [15]:
list(df.columns)

['sr_no',
 'patient_id',
 'pcos',
 'age_yrs',
 'weight_kg',
 'height_cm',
 'bmi',
 'blood_group',
 'pulse_rate',
 'respiratory_rate',
 'hb',
 'cycle_type',
 'cycle_length',
 'marriage_yrs',
 'pregnant',
 'No. of aborptions',
 'I   beta-HCG(mIU/mL)',
 'II    beta-HCG(mIU/mL)',
 'fsh',
 'lh',
 'fsh_lh_ratio',
 'hip_inch',
 'waist_inch',
 'waist_hip_ratio',
 'tsh',
 'amh',
 'prolactin',
 'vit_d3',
 'progesterone',
 'rbs',
 'weight_gain',
 'hair_growth',
 'skin_darkening',
 'hair_loss',
 'pimples',
 'fast_food',
 'exercise',
 'bp_sys',
 'bp_dia',
 'follicle_left',
 'follicle_right',
 'follicle_size_left',
 'follicle_size_right',
 'endometrium']

In [16]:
df = df.rename(columns={
    "No. of abortions": "abortions",
    "I   beta-HCG(mIU/mL)": "beta_hcg_1",
    "II    beta-HCG(mIU/mL)": "beta_hcg_2"
})

## Fix Data Types

In [17]:
df["bmi"] = pd.to_numeric(df["bmi"], errors="coerce")
df["beta_hcg_2"] = pd.to_numeric(df["beta_hcg_2"], errors="coerce")

## Missing Value Analysis

In [18]:
df.isnull().sum().sort_values(ascending=False)

bmi                    1
fast_food              1
beta_hcg_2             1
marriage_yrs           1
age_yrs                0
weight_kg              0
height_cm              0
blood_group            0
pulse_rate             0
sr_no                  0
pcos                   0
patient_id             0
cycle_type             0
hb                     0
respiratory_rate       0
pregnant               0
No. of aborptions      0
beta_hcg_1             0
fsh                    0
cycle_length           0
fsh_lh_ratio           0
hip_inch               0
waist_inch             0
waist_hip_ratio        0
tsh                    0
amh                    0
prolactin              0
lh                     0
vit_d3                 0
progesterone           0
weight_gain            0
rbs                    0
hair_growth            0
skin_darkening         0
hair_loss              0
pimples                0
exercise               0
bp_sys                 0
bp_dia                 0
follicle_left          0


## Duplicate Records

In [19]:
df.duplicated().sum()

np.int64(0)

## PCOS Distribution

In [20]:
df["pcos"].value_counts()

pcos
0.0    364
1.0    176
Name: count, dtype: int64

## Thus, Acc. To Data
No PCOS → 364 patient
PCOS present → 176 patients

## Statistical Summary

In [21]:
df.describe()

,sr_no,patient_id,pcos,age_yrs,weight_kg,height_cm,bmi,blood_group,pulse_rate,respiratory_rate,...,pimples,fast_food,exercise,bp_sys,bp_dia,follicle_left,follicle_right,follicle_size_left,follicle_size_right,endometrium
count,540.000000,540.000000,540.000000,540.000000,540.000000,540.000000,539.000000,540.000000,540.000000,540.000000,...,540.000000,539.000000,540.000000,540.000000,540.000000,540.000000,540.000000,540.000000,540.000000,540.000000
mean,271.274074,271.274074,0.325926,31.446296,59.614259,156.469111,24.317792,13.803704,73.250000,19.246296,...,0.490741,0.513915,0.248148,114.651852,76.940741,6.125926,6.635185,15.016296,15.446981,8.474389
std,156.332298,156.332298,0.469154,5.403813,11.025637,6.027976,4.057048,1.842194,4.434067,1.689344,...,0.500378,0.500271,0.432339,7.387819,5.571279,4.232448,4.438642,3.569895,3.320108,2.167098
min,1.000000,1.000000,0.000000,20.000000,31.000000,137.000000,12.420000,11.000000,13.000000,16.000000,...,0.000000,0.000000,0.000000,12.000000,8.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,136.750000,136.750000,0.000000,28.000000,52.000000,152.000000,21.675000,13.000000,72.000000,18.000000,...,0.000000,0.000000,0.000000,110.000000,70.000000,3.000000,3.000000,13.000000,13.000000,7.000000
50%,271.500000,271.500000,0.000000,31.000000,59.000000,156.000000,24.240000,14.000000,72.000000,18.000000,...,0.000000,1.000000,0.000000,110.000000,80.000000,5.000000,6.000000,15.000000,16.000000,8.500000
75%,406.250000,406.250000,1.000000,35.000000,65.000000,160.000000,26.655000,15.000000,74.000000,20.000000,...,1.000000,1.000000,0.000000,120.000000,80.000000,9.000000,10.000000,18.000000,18.000000,9.800000
max,541.000000,541.000000,1.000000,48.000000,108.000000,180.000000,38.900000,18.000000,82.000000,28.000000,...,1.000000,1.000000,1.000000,140.000000,100.000000,22.000000,20.000000,24.000000,24.000000,18.000000


### Observation

The dataset contains 540 patient records with clinical, hormonal, and lifestyle-related features associated with PCOS prediction.

Key observations from the statistical summary:

- The average age of patients is around 31 years.
- Average BMI is approximately 24, indicating most patients fall near the normal to overweight range.
- Weight ranges from 31 kg to 108 kg, showing significant variation among individuals.
- Height ranges between 137 cm and 180 cm.
- PCOS values are binary (0 = No PCOS, 1 = PCOS), confirming this is a classification problem.
- Several healthcare measurements such as hormone levels, blood pressure, and follicle counts show wide variability, which may influence prediction performance.

The statistical summary helps understand feature distribution, detect unusual values, and prepare the dataset for exploratory analysis and machine learning.
